ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Extended.hs

```diff
@@ -81,7 +88,7 @@ foldExtLedgerState :: (ProtocolLedgerView blk, HasCallStack)
                    -> [blk] -- ^ Blocks to apply, oldest first
                    -> ExtLedgerState blk
                    -> Except (ExtValidationError blk) (ExtLedgerState blk)
-foldExtLedgerState = repeatedlyM . applyExtLedgerState
+foldExtLedgerState = repeatedlyM . (applyExtLedgerState False)
```

Comment:
Should probably leave a comment in the code explaining this hardcoded `False`:

It doesn't seem like it would make much sense to be able to flag blocks as "previously applied" (or "known to be valid") and re-apply them in a fold operation since a block should only be re-applied to a single specific ledger state in which we know the block to be valid. There would be no way to know if a subsequent block is valid with respect to the resulting ledger state of the last `applyExtLedgerState` operation.

Reply:
Yeah, best to add a comment. This function is only used in the tests, so the potential performance loss here is not important.

I think it *could* make sense to flag a bunch of blocks as "previously applied". For instance when restarting the ChainDB: you might have to replay a bunch of blocks from the ImmutableDB (so they have been "previously applied" and *must* be valid) against an older snapshot of the ledger (a recent one might be missing or corrupt). In this case, this would be a welcome speed-up of the start-up. Note that we are already *reapplying* in this scenario, but we are just not using `foldExtLedgerState` for it, see: https://github.com/input-output-hk/ouroboros-network/blob/aef9ef24cee0ecf28588fa61def7ffc1e9d75838/ouroboros-consensus/src/Ouroboros/Storage/LedgerDB/OnDisk.hs#L236

Another (more contrived) example: say you want to do chain selection on a bunch of chains. First, you have to validate these chains, but you don't want to keep a ledger state in memory for each chain, because that would take too much memory. So you after validating each chain, you throw the resulting ledger state away, then choose a valid chain, and for that chain, you quickly reapply the blocks to obtain your final ledger state snapshot. Note that this was once how we did chain selection, but we do it differently now without requiring us to validate all candidate chains up front.

Reply:
@mrBliss: thanks, that's a good point!

Reply:
I'm just adding a param for specifying whether the blocks have been previously applied.

Reply:
> @mrBliss: I would like to add some tests for these changes but I think that should come in another PR. We'll need to add some generators for valid Byron blocks, transactions, etc.
> 
> edit: I believe that some related generators are being worked on in #855?

Indeed. Maybe you can assist @ruhatch with that? Can you make a ticket for these tests?